gnocchi/chef.py

```diff
@@ -45,9 +45,104 @@ class Chef(object):
     def __init__(self, coord, incoming, index, storage):
         self.coord = coord
         self.incoming = incoming
+        # This variable is an instance of the indexer,
+        # which means, database connector.
         self.index = index
         self.storage = storage
 
+    def clean_raw_data_inactive_metrics(self):
+        """Cleans metrics raw data if they are inactive.
+
+        The truncating of the raw metrics data is only done when new
+        measures are pushed. Therefore, if no new measures are pushed, and the
+        archive policy was updated to reduce the backwindow, the raw
+        datapoints for metrics that are not receiving new datapoints are never
+        truncated.
+
+        The goal of this method is to identify metrics that are in
+        "inactive state", meaning, not receiving new datapoints, and execute
+        their raw data points truncation. We check the column
+        "needs_raw_data_truncation", to determine if the archive policy was
+        updated, and no measure push was executed for the metric.
+
+        If the metric is not receiving new datapoints, the processing workflow
+        will not mark the column "needs_raw_data_truncation" to False;
+        therefore, that is how we identify such metrics.
+        """
+
+        metrics_to_clean = self.index.list_metrics(
+            attribute_filter={"==": {
+                "needs_raw_data_truncation": True}}
+        )
+
+        LOG.debug("Metrics [%s] found to execute the raw data cleanup.",
+                  metrics_to_clean)
+
+        for metric in metrics_to_clean:
```

Comment:
here you are listing the metrics to clean, iterate over them, calculating each metric sack, locking the sack, processing the metric and unlocking the sack, so with many metrics in a sack, means lots of lock and unlocking requests and increasing the number of lock collisions.

Wouldn't be better, if we:

1.  List the metrics to cleanup;
2. Calculate their sacks and group they by sacks (like map sack -> list of metrics);
3. Iterate over the sacks, lock the sack
4. Iterate over metrics in the sack and process the metric;
5. unlock the sack when the metrics processing finishes;

Then you would lock n sacks instead of lock n metrics.

Reply:
Done. Can you check if the solution is fine by you now?